hydradx/model/amm/centralized_market.py

```diff
@@ -0,0 +1,254 @@
+                        tkns_bought += bid[0] * sell_tkns_remaining
+                        bid[1] -= sell_tkns_remaining
+                        sell_tkns_remaining = 0
+                    else:
+                        # this bid can partially fill the order
+                        tkns_bought += bid[1] * bid[0]
+                        sell_tkns_remaining -= bid[1]
+                        bid[1] = 0
+                    if bid[1] == 0:
+                        remove_bids += 1
+                    if sell_tkns_remaining <= 0:
+                        break
+            else:
+                for ask in self.order_book[(base, quote)].asks:
+                    if ask[0] * ask[1] >= sell_tkns_remaining:
+                        tkns_bought += sell_tkns_remaining / ask[0]
+                        ask[1] -= sell_tkns_remaining / ask[0]
+                        sell_tkns_remaining = 0
+                    else:
+                        tkns_bought += ask[1]
+                        sell_tkns_remaining -= ask[1] * ask[0]
+                        ask[1] = 0
+                    if ask[1] == 0:
+                        remove_asks += 1
+                    if sell_tkns_remaining <= 0:
+                        break
+
+            agent.holdings[tkn_sell] -= sell_quantity - sell_tkns_remaining
+            agent.holdings[tkn_buy] += tkns_bought * (1 - self.trade_fee)
+
+        elif buy_quantity > 0:
+            buy_tkns_remaining = buy_quantity
+            tkns_sold = 0
+
+            if tkn_buy == base:
+                for ask in self.order_book[(base, quote)].asks:
+                    if ask[1] >= buy_tkns_remaining:
+                        tkns_sold += buy_tkns_remaining * ask[0]
+                        ask[1] -= buy_tkns_remaining
+                        buy_tkns_remaining = 0
+                    else:
+                        tkns_sold += ask[0] * ask[1]
+                        buy_tkns_remaining -= ask[1]
+                        ask[1] = 0
+                    if ask[1] == 0:
+                        remove_asks += 1
+                    if buy_tkns_remaining <= 0:
+                        break
+            else:
+                for bid in self.order_book[(base, quote)].bids:
+                    if bid[0] * bid[1] >= buy_tkns_remaining:
+                        tkns_sold += buy_tkns_remaining / bid[0]
+                        bid[1] -= buy_tkns_remaining / bid[0]
+                        buy_tkns_remaining = 0
+                    else:
+                        tkns_sold += bid[1]
+                        buy_tkns_remaining -= bid[0] * bid[1]
+                        bid[1] = 0
+                    if bid[1] == 0:
+                        remove_bids += 1
+                    if buy_tkns_remaining <= 0:
+                        break
+
+            agent.holdings[tkn_buy] += buy_quantity - buy_tkns_remaining
+            agent.holdings[tkn_sell] -= tkns_sold / (1 - self.trade_fee)
```

Comment:
I think this should be tkns_sold * (1 + trade_fee), because Kraken says they always charge fee in quote asset. The way that would work I'd expect is that if the fee is e.g. 1% and I am buying TKN with QTE, I pay 1% extra QTE.